### ISSUE TITLE
[Feature request] Implement speedtest

### ISSUE THREAD

**Issue by kkoomen** [mullvad/mullvadvpn-app#1113]
**Background info:**
I currently reside in China for a long time and since the 2019.06 (or 2019.07) with the stable Wireguard support it is by far the fastest VPN so far, and yes it beats ExpressVPN by far (yay for that!).

**Feature request: speedtest + auto-connect**
Since the wireguard servers their speed different quite a lot every moment of the day I did made my own (Python 3) script which does speedtests for all the wireguard servers and it will provide me back with a top-10 result and then it will also automatically connect to the best server it checked with the fastest download speed (or maybe best download/upload ratio, but I am more interested in download, as most people are).

I really like to see this feature back in the app. Even though I do not enjoy ExpressVPN, their speedtest works great. It still keeps you connected to your current location and is still able to speedtest for you. I have no idea how they do that, but it might be an inspirational source.

**Improved feature request: speedtest + auto-connect + auto-multihop**
An improved version of my local speedtest script I want to add is that the first server will become the main connection and then the second-best one will be the multihop server automatically. It would be great if this could be a feature as well.

**Continues async background:**
If possible, I'd also prefer this to be something in the async in the background. This would ofcourse be logical, but with "async" I more refer to that it should be possible to do speedtests while I can still continue with browsing the internet on my current connection. This is something my script can't do unfortunately and as I mentioned: ExpressVPN is able to do this as well iirc.

I am also referring to that (maybe this can be the main functionality) to always run speedtests and then have an indicator next to the title per location in the locations-overview that indicates how fast it is compared to other servers.

**Sorting by speed**
If this feature will be implemented I also expect an option for me to choose how I want to sort my locations: by name or speed or something else.

<hr />

Since server speeds might differ a lot during the day (at least here), I'd like Mullvad to check these things for me with a single press on a button and then it will do the rest.

I'd like to get feedback as well whether this idea is possible and to what extend this is possible or different suggestions on how to do things.

Thanks.

**Comment by firecodeing** [comment]
I think this could be a really big improvement. That way Mullvad will be a much more solid provider in country's where they don't want to use a VPN.

**Comment by kayg04** [comment]
Yes, this would be great but seeing this is something that generates a lot of noise, making this feature a toggle-able option under settings -> preferences would be really nice.

**Comment by gudh** [comment]
Speed test is a must have!

Btw, could you share your Python3 speed test script if possible? Or some hints on how to write one?

Thanks in advance!

**Comment by faern** [comment]
> Btw, could you share your Python3 speed test script if possible? Or some hints on how to write one?

I don't know what script you are talking about. We don't have one. Can't you use `speedtest-cli` or something like that to get an estimate for your bandwidth?

**Comment by gudh** [comment]
@faern 

Sorry I didn't make things clear, 

Because I saw @kkoomen mentioned that he wrote a Python3 script to do the speed test thing, I wonder if he could share it since speed test is a frequent used feature for me back when I was using other VPNs

**Comment by hbednar** [comment]
Something like this would be really helpful and may help to improve performance.

![image](https://user-images.githubusercontent.com/71215137/222092982-e3dfe967-0aa4-443c-8e21-2835494d3208.png)

**Comment by inson1** [comment]
It would be really hepfull and it could help us with managing your resources better
